app/src/main/java/com/duckduckgo/app/tabs/model/TabDataRepository.kt

```diff
@@ -137,6 +158,23 @@ class TabDataRepository @Inject constructor(
         siteData.remove(tab.tabId)
     }
 
+    override suspend fun deleteCurrentTabAndSelectSource() {
+        databaseExecutor().scheduleDirect {
+            val tabToDelete = tabsDao.selectedTab()
+            tabToDelete?.tabId?.let {
+                deleteOldPreviewImages(tabToDelete.tabId)
+                tabsDao.deleteTab(tabToDelete)
+                siteData.remove(tabToDelete.tabId)
+            }
+
+            tabToDelete?.sourceTabId?.let { sourceTabId ->
```

Comment:
Checking twice this method, I was asking myself what will happen if `tabToDelete?.sourceTabId` is null. It will not crash but we will not select a new tab. It may produce an inconsistent state for users.
Additionally, we are doing a few operations that modify the database (delete + insert), but they are not inside a transaction. If one fails, we could leave the database in a weird state too.

To solve both things I was thinking about modifying `tabsDao.deleteTabAndUpdateSelection`, where we can have a `@Transaction`, and also a fallback to the firstTab.

```
    @Transaction
    open fun deleteTabAndUpdateSelection(tab: TabEntity, newSelectedTab: TabEntity? = null) {
        deleteTab(tab)

        if (newSelectedTab != null) {
            insertTabSelection(TabSelectionEntity(tabId = newSelectedTab.tabId))
            return
        }

        if (selectedTab() != null) {
            return
        }

        firstTab()?.let {
            insertTabSelection(TabSelectionEntity(tabId = it.tabId))
        }
    }
```

and the repository method will look like:
```
    override suspend fun deleteCurrentTabAndSelectSource() {
        databaseExecutor().scheduleDirect {
            val tabToDelete = tabsDao.selectedTab() ?: return@scheduleDirect

            deleteOldPreviewImages(tabToDelete.tabId)
            val tabToSelect = tabToDelete.sourceTabId
                .takeUnless { it.isNullOrBlank() }
                ?.let {
                    tabsDao.tab(it)
                }
            tabsDao.deleteTabAndUpdateSelection(tabToDelete, tabToSelect)
            siteData.remove(tabToDelete.tabId)
        }
    }
```

I haven't checked if that actually works or if it compiles 😅, just wanted to show the idea. What do you think? please double check from your side too in case you like the approach.

Reply:
Good call, would definitely want to ensure the entire transaction completes rather than risk leaving the DB in an undesired state. The code you provided works as is and looks good to me.

This also accounts for the unlikely situation that a user opens a link in the new tab, declines the dialog, closes the sourceTab, and then refreshes the download tab to reinitialize the download dialog and proceeds. The user will now return to the firstTab where as previously the user would just remain on the blank new tab.